File: Hub/src/Core/Devices/Device.cs
using static Core.Devices.MeasuredDataPoint;

namespace Core.Devices
{

    public abstract class Device
    {
        public enum DeviceType
        {
            TemperatureSensor = 0x0,
            SoilMoistureSensor = 0x1,
            SoilTemperatureSensor = 0x2,
            WaterPumpDevice = 0x3,
        }

        public DeviceType Type { get; private set; }
        public abstract uint State { get; }
        public byte DeviceId { get; private set; }
        public uint UniqueId { get; private set; }
        public string DeviceName { get; private set; }
        private Dictionary<DataPointType, MeasuredDataPoint> _lastMeasuredDataPoints = new Dictionary<DataPointType, MeasuredDataPoint>();
        public IReadOnlyList<MeasuredDataPoint> LastMeasuredDataPoints => _lastMeasuredDataPoints.Values.ToList().AsReadOnly();

        public Device(DeviceType type, byte deviceId, uint uniqueId, string deviceName)
        {
            Type = type;
            DeviceId = deviceId;
            UniqueId = uniqueId;
            DeviceName = deviceName;
        }

        public void SetDeviceName(string newName)
        {
            if (string.IsNullOrEmpty(newName))
            {
                throw new Exception("Name is empty");
            }

            DeviceName = newName;
        }

        protected abstract MeasuredDataPoint HandleDataSource(int value, byte dataSource);

        public void AddMeasurementDataPoint(int value, byte dataSource)
        {
            var dataPoint = HandleDataSource(value, dataSource);
            if (_lastMeasuredDataPoints.ContainsKey(dataPoint.Type))
            {
                _lastMeasuredDataPoints[dataPoint.Type] = dataPoint;
            }
            else
            {
                _lastMeasuredDataPoints.Add(dataPoint.Type, dataPoint);
            }
        }

        public abstract List<byte> GetDataSourceTypes();
    }
}

File: Hub/src/WebAPI/Client/src/models/Device.ts
export enum DeviceType {
    TemperatureSensor = 0,
    SoilMoistureSensor = 1,
    SoilTemperatureSensor = 2,
    WaterPumpDevice = 3,
}

export enum SensorDeviceState {
    Normal = 0,
    Error = 1,
    Unresponsive = 2,
}

export enum ControlDeviceState {
    Idle = 0,
    WaitingForResponse = 1,
    PrimaryActionRunning = 2,
}

export enum DataPointType {
    Temperature = 0,
    SoilMoisture = 1,
    Humidity,
    WaterLevel
}

export type MeasuredDataPoint = {
    type: DataPointType;
    value: number;
    dateTime: Date;
}

export type Device = {
    type: DeviceType;
    deviceId: number;
    deviceName: string;
    state: number;
    lastMeasuredDataPoints: MeasuredDataPoint[];
}